feat: Add support for reading unit cell dimensions from MOL2 CRYSIN records

Dreamstick9 opened the pull request:
Fixes #3341

Changes made in this Pull Request:

Implemented logic in MOL2Parser (package/MDAnalysis/coordinates/MOL2.py) to parse @<TRIPOS>CRYSIN records.

Extracted unit cell dimensions (lengths and angles) from the CRYSIN block and assigned them to ts.dimensions.

Added a new test data file with a CRYSIN record: testsuite/MDAnalysisTests/data/mol2/test_crysin.mol2.

Registered the new test file in testsuite/MDAnalysisTests/datafiles.py.

Added a regression test test_mol2_crysin_dimensions in testsuite/MDAnalysisTests/coordinates/test_mol2.py to verify dimensions are loaded correctly.

PR Checklist
[x] Issue raised/referenced?

[x] Tests updated/added?

[ ] Documentation updated/added?

[x] package/CHANGELOG file updated?

[x] Is your name in package/AUTHORS? (If it is not, add it!)

Developers Certificate of Origin
I certify that I can submit this code contribution as described in the [Developer Certificate of Origin](https://developercertificate.org/), under the MDAnalysis [LICENSE](https://github.com/MDAnalysis/mdanalysis/blob/develop/LICENSE).


<!-- readthedocs-preview mdanalysis start -->
----
 Documentation preview : https://mdanalysis--5187.org.readthedocs.build/en/5187/

<!-- readthedocs-preview mdanalysis end -->

Dreamstick9 committed:
feat: enable reading unit cell dimensions from MOL2 CRYSIN records

Dreamstick9 committed:
test: add test case for MOL2 CRYSIN unit cell dimensions

Dreamstick9 committed:
style: apply black formatting

Dreamstick9 committed:
fix: add mol2_crysin to __all__ exports and format

Dreamstick9 committed:
style: run black v24 on all files

Dreamstick9 committed:
style: reformat MOL2.py to satisfy CI

Dreamstick9 committed:
style: black formatting for tests and datafiles

orbeckst commented:
@Dreamstick9 can you link to the document that describes the `@CRYSIN` record? Can you show/link to examples mol2 files found "in the wild" that use it, e.g., from a database?

That may help with reviewing.

Also make sure that you address all points in the checklist (eg, add a note to the docs, add an entry to CHANGELOG). Reviewing PRs takes a lot of time and time is really precious so show reviewers that you've done all the necessary work.

Dreamstick9 committed:
docs: update changelog, authors, and parser docstrings

Dreamstick9 commented:
I implemented  this based on the official Tripos mol2 file format specification, the @crysin record is defined the the standard manual in page 17 of the pdf(https://github.com/UnixJunkie/mol2-file-format-spec/blob/master/mol2.pdf). the test case added in mol2_crysin was for this specification. this record is the standard method for defining periodic boundary conditions in the format (see "Data Record Format" in the spec).
i have updated the CHANGELOG and AUTHORS.

Thanks for the review @orbeckst

Dreamstick9 committed:
black formatting

orbeckst commented on package/CHANGELOG:
```diff
@@ -230,6 +230,7 @@ Fixes
  * Fix doctest errors of lib/picklable_file_io.py (Issue #3925, PR #4371)
  * Fix deploy action to use the correct version of the pypi upload action.
  * Fix groups.py doctests using sphinx directives (Issue #3925, PR #4374)
+ * MOL2Parser now reads unit cell dimensions from @<TRIPOS>CRYSIN records (Issue #3341)
```
This line needs to go under Enhancements for 2.11.0 (near line 39).

 add your Github handle `Dreamstick9` into the list of contributors on line 9.

orbeckst commented on package/MDAnalysis/coordinates/MOL2.py:
```diff
@@ -246,6 +246,13 @@ def _read_frame(self, frame):
+                dims = [float(x) for x in line.split()[:6]]
+                self.ts.dimensions = np.array(dims, dtype=np.float32)
+            except (ValueError, IndexError):
+                pass
```
Under which circumstances will the exception be triggered?

Do we need the try/except?

There's currently no test for it so if we need it then it should get tested.

orbeckst reviewed (changes requested):
I have some inline comments/questions — please address those.

Update the docs where it says that MOL2 ignores CRYSIN to say that it ignores the spacegroup and settings. The unitcell is interpreted in the common crystallographic convention with box vector **a** parallel to X, **b** in the XY plane, and **c** with a positive Z component.

orbeckst commented on package/MDAnalysis/coordinates/MOL2.py:
```diff
@@ -246,6 +246,13 @@ def _read_frame(self, frame):
+        if "crysin" in sections:
+            try:
+                line = sections["crysin"][0].strip()
+                dims = [float(x) for x in line.split()[:6]]
```
Note: I spent some time figuring out if we need spacegroup and setting (see https://github.com/MDAnalysis/mdanalysis/issues/3341#issuecomment-3770093270) and came to the conclusion that we can ignore thm. However, we should document that — see my review comment.

Dreamstick9 committed:
feat: add support for reading unit cell dimensions from @<TRIPOS>CRYSIN records in MOL2Parser

Dreamstick9 committed:
Merge branch 'develop' into feat/3341-add-crysin-support

Dreamstick9 commented:
Thanks for the review @orbeckst,  I've made the requested changes and also removed the try and except,

Dreamstick9 committed:
Merge branch 'develop' into feat/3341-add-crysin-support

Dreamstick9 committed:
Merge branch 'develop' into feat/3341-add-crysin-support

tylerjereddy commented on package/CHANGELOG:
```diff
@@ -15,7 +15,7 @@ The rules for this file:
 -------------------------------------------------------------------------------
 ??/??/?? IAlibay, orbeckst, marinegor, tylerjereddy, ljwoods2, marinegor,
-         spyke7, talagayev, tanii1125, BradyAJohnston, hejamu
+         spyke7, talagayev, Dreamstick9
```
looks like Brady and a few others got accidentally deleted from the author list here?

tylerjereddy commented on testsuite/MDAnalysisTests/coordinates/test_mol2.py:
```diff
@@ -232,3 +234,11 @@ def test_mol2_universe_write(tmpdir):
+    u = mda.Universe(mol2_crysin)
+
+    expected = np.array([40.0, 50.0, 60.0, 90.0, 90.0, 90.0], dtype=np.float32)
+    assert_array_almost_equal(u.dimensions, expected, decimal=3)
```
`assert_allclose()` is generally preferable in new testing code (see docs of `assert_array_almost_equal`)

Dreamstick9 committed:
fix: restore CHANGELOG authors and use assert_allclose per review

Dreamstick9 commented:
@tylerjereddy  i've updated the CHANGELOG and also switched to `assert_allclose` in `test_mol2.py`
Thanks for the review!

Dreamstick9 committed:
Merge branch 'develop' into feat/3341-add-crysin-support

Dreamstick9 committed:
Update CHANGELOG

orbeckst committed:
Merge branch 'develop' into feat/3341-add-crysin-support

orbeckst commented on package/MDAnalysis/coordinates/MOL2.py:
```diff
@@ -69,8 +69,11 @@
   currently read or preserved on writing.
-* As the CRYSIN record type is not parsed / written, MOL2 systems always have
-  dimensions set to ``None`` and dimensionless MOL2 files are written.
+* The MDAnalysis :class:`MOL2Reader` now reads unit cell dimensions from the
```
```suggestion
* The MDAnalysis :class:`MOL2Reader` reads unit cell dimensions from the
```

orbeckst commented on package/MDAnalysis/coordinates/MOL2.py:
```diff
@@ -246,6 +249,10 @@ def _read_frame(self, frame):
+        if "crysin" in sections:
+            line = sections["crysin"][0].strip()
+            dims = [float(x) for x in line.split()[:6]]
+            self.ts.dimensions = np.array(dims, dtype=np.float32)
```
You should be able to have np do the cast (which is a _tiny_ bit faster)
```suggestion
            dims = line.split()[:6]
            self.ts.dimensions = np.array(dims, dtype=np.float32)
```

orbeckst reviewed (approved):
Thank you for addressing my comments. I have two changes that I am going to make and then it's good to go from my end.

orbeckst committed:
Apply suggestions from code review

orbeckst commented:
Congratulations @Dreamstick9 , your contribution is merged 🎉 ! Thank you!
